_posts/2024-05-14-registry-dereference.md

```diff
@@ -0,0 +1,285 @@
+          },
+          "name" : {
+            "description" : "The subject's name",
+            "type" : "string"
+          }
+        }
+      }
+    },
+    "city" : {
+      "title" : "City",
+      "type" : "object",
+      "properties" : {
+        "zipCode" : {
+          "description" : "The zip code.",
+          "type" : "integer",
+          "minimum" : 0
+        },
+        "qualification" : {
+          "title" : "Qualification",
+          "type" : "object",
+          "properties" : {
+            "qualification" : {
+              "description" : "The city qualification",
+              "type" : "integer",
+              "minimum" : 10
+            },
+            "name" : {
+              "description" : "The subject's name",
+              "type" : "string"
+            }
+          }
+        },
+        "name" : {
+          "description" : "The city's name.",
+          "type" : "string"
+        }
+      }
+    },
+    "age" : {
+      "description" : "Age in years which must be equal to or greater than zero.",
+      "type" : "integer",
+      "minimum" : 0
+    }
+  },
+  "required" : [ "city" ],
+  "$id" : "https://example.com/citizen.json"
+}
+```
+As you can see, that's the content of the full hierarchy, not just the schema of the _citizen_ element alone. There are two endpoints where you can get the content dereferenced, with the following curl commands:
+
+```
+curl --location 'http://localhost:8080/apis/registry/v2/ids/globalIds/7?dereference=true'
+curl --location 'http://localhost:8080/apis/registry/v2/groups/default/artifacts/JsonSerdeReferencesExample?dereference=true'
+```
+
+JSON Schema Kafka Serde dereference
+===
+
+Now, in order to make good use of all of this, we can use a Kafka producer. For this example, in the Kafka Producer there is one extremely important property we must set, the schema resolver configuration option _SERIALIZER_DEREFERENCE_SCHEMA_. This is the option that will instruct the Schema Resolver to use the dereference parameter in the globalIds endpoint above (since that is the endpoint used by default to resolve the schema).
+
+```java
+Properties props = new Properties();
+
+// Configure kafka settings
+props.putIfAbsent(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, SERVERS);
+props.putIfAbsent(ProducerConfig.CLIENT_ID_CONFIG, "Producer-" + TOPIC_NAME);
+props.putIfAbsent(ProducerConfig.ACKS_CONFIG, "all");
+props.putIfAbsent(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
+props.putIfAbsent(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, JsonSchemaKafkaSerializer.class.getName());
+props.putIfAbsent(SerdeConfig.ARTIFACT_RESOLVER_STRATEGY, SimpleTopicIdStrategy.class.getName());
+props.putIfAbsent(SerdeConfig.EXPLICIT_ARTIFACT_GROUP_ID, "default");
+props.putIfAbsent(SerdeConfig.REGISTRY_URL, REGISTRY_URL);
+
+//This is the most important configuration for this particular example.
+props.putIfAbsent(SchemaResolverConfig.SERIALIZER_DEREFERENCE_SCHEMA, "true");
```

Comment:
What does this do for the producer/serializer?

Reply:
In the case of JSON Schema, the schema is already registered, so when it's fetched to validate the data before it's being produced, it gets the schema dereferenced. For avro, if it's enabled and registration is enabled as well, it registers the schema dereferenced (since the schema is inferred from the data, not determined by the user).